### PR TITLE
Make expiry safe from racing against entry dedupe

### PR DESCRIPTION
**Pull request by arielshaqed**
See https://docs.google.com/document/d/1z9HQpkNtgvnx7y8y85N-w3a_mj6ZoOeE9KBdrsfw6TM/edit#heading=h.ru8lo6tlexv4

### CODE REVIEW

File: cmd/lakefs/cmd/expire.go [hunk 1]
```diff
@@ -84,7 +85,8 @@ var expireCmd = &cobra.Command{
 				// (not a failure)
 				continue
 			}
-			expiryRows, err := cataloger.QueryExpired(ctx, repo.Name, &policy.Policy)
+			// TODO(ariels): Rewrite this!
```
Comment (ozkatz) [inline cmd/lakefs/cmd/expire.go:88]: I suggest avoiding TODO comments and open a story instead.. these are action items that aren't tracked.

Comment (ozkatz) [reply cmd/lakefs/cmd/expire.go:88]: also, why does it require rewriting?

File: catalog/cataloger_retention.go [hunk 2]
```diff
@@ -179,14 +195,18 @@ func (e *expiryRows) Read() (*ExpireResult, error) {
 	}, nil
 }
 
-func (c *cataloger) QueryExpired(ctx context.Context, repositoryName string, policy *Policy) (ExpiryRows, error) {
+func (c *cataloger) QueryEntriesToExpire(ctx context.Context, repositoryName string, policy *Policy) (ExpiryRows, error) {
 	logger := logging.FromContext(ctx).WithField("policy", *policy)
 
+	// TODO(ariels): page!
```
Comment (ozkatz) [inline catalog/cataloger_retention.go:201]: see my other note on TODO comments

File: catalog/cataloger_retention.go [hunk 3]
```diff
@@ -274,3 +294,65 @@ func (c *cataloger) MarkExpired(ctx context.Context, repositoryName string, expi
 	logger.WithField("count", count).Info("expired records")
 	return nil
 }
+
+// TODO(ariels): chunk.
+func (c *cataloger) MarkObjectsForDeletion(ctx context.Context, repositoryName string) (int64, error) {
+	// TODO(ariels): This query is difficult to chunk.  One way: Perform the inner SELECT
```
Comment (ozkatz) [inline catalog/cataloger_retention.go:300]: TODO comment, see note.

File: catalog/cataloger_retention.go [hunk 4]
```diff
@@ -274,3 +294,65 @@ func (c *cataloger) MarkExpired(ctx context.Context, repositoryName string, expi
...
+	return s.rows.Close()
+}
+
+func (s *StringRows) Read() (string, error) {
+	var ret string
+	err := s.rows.Scan(&ret)
+	return ret, err
+}
+
+// TODO(ariels): Process in chunks.  Can store the inner physical_address query in a table for
```
Comment (ozkatz) [inline catalog/cataloger_retention.go:338]: TODO comment, see note.

File: retention/expiry.go [hunk 5]
```diff
@@ -77,35 +77,30 @@ func (bw *BucketWriters) GetWriter(bucketName string) (*csv.Writer, error) {
...
-func WriteExpiryManifestsFromReader(ctx context.Context, c catalog.Cataloger, r io.Reader) (map[string]fileutil.RewindableReader, error) {
+func WriteExpiryManifestsFromRows(ctx context.Context, c catalog.Cataloger, repository *catalog.Repository, rows catalog.StringRows) (map[string]fileutil.RewindableReader, error) {
 	logger := logging.FromContext(ctx)
-	decoder := json.NewDecoder(r)
 	bucketWriters := BucketWriters{}
 
 	var err error
-	for recordNumber := 0; ; recordNumber++ {
+	recordNumber := 0
+	for ; rows.Next(); recordNumber++ {
```
Comment (ozkatz) [inline retention/expiry.go:86]:
```suggestion
	for recordNumer := 0; rows.Next(); recordNumber++ {
```
?

File: retention/expiry.go [hunk 6]
```diff
@@ -250,21 +287,93 @@ type ExpireOnS3Params struct {
...
+	}
 }
 
 // ExpireOnS3 starts a goroutine to expire all entries on expiryResultsReader and returns a
 // channel that will receive all error results.
-func ExpireOnS3(ctx context.Context, s3ControlClient s3controliface.S3ControlAPI, s3Client s3iface.S3API, c catalog.Cataloger, expiryResultsReader fileutil.RewindableReader, params *ExpireOnS3Params) chan error {
+func ExpireOnS3(ctx context.Context, s3ControlClient s3controliface.S3ControlAPI, s3Client s3iface.S3API, c catalog.Cataloger, repository *catalog.Repository, expiryResultsReader fileutil.RewindableReader, params *ExpireOnS3Params) chan error {
+	const markExpiredChunkSize = 50000
+
+	// TODO(ariels): Lock something for this "session" (e.g. process)
```
Comment (ozkatz) [inline retention/expiry.go:330]: see note on TODO comments

File: retention/expiry.go (hunk 5 above)
Comment (ozkatz) [reply retention/expiry.go:86]: :+1: missed the usage outside the loop.

**Comment by arielshaqed** [comment]
Thanks!

**Pull request merged**